src/rust/src/wallet.rs

```diff
@@ -130,38 +151,164 @@ pub enum WalletError {
+        self.last_checkpoint = Some(block_height);
+        true
+    }
+
+    /// Returns whether or not a checkpoint has been created. If no checkpoint exists,
+    /// the wallet has not yet observed any blocks.
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    /// Rewinds the note commitment tree to the given height, removes notes and
+    /// spentness information for transactions mined in the removed blocks, and returns
+    /// the number of blocks by which the tree has been rewound if successful. Returns  
+    /// `RewindError` if not enough checkpoints exist to execute the full rewind
+    /// requested. If the requested height is greater than or equal to the height of the
+    /// latest checkpoint, this returns `Ok(0)` and leaves the wallet unmodified.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> Result<u32, RewindError> {
+        if let Some(checkpoint_height) = self.last_checkpoint {
+            if to_height >= checkpoint_height {
+                return Ok(0);
+            }
+
+            let blocks_to_rewind = <u32>::from(checkpoint_height) - <u32>::from(to_height);
+
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let checkpoint_count = self.witness_tree.checkpoints().len();
+            if blocks_to_rewind as usize > checkpoint_count {
+                return Err(RewindError::InsufficientCheckpoints(checkpoint_count));
+            }
+
+            let mut blocks_rewound: u32 = 0;
+            while blocks_rewound < blocks_to_rewind {
+                // we've already checked that we have enough checkpoints to fully
+                // rewind by the requested amount.
+                assert!(self.witness_tree.rewind());
+                blocks_rewound += 1;
+            }
+
+            // reset our last observed height to ensure that notes added in the future are
+            // from a new block
+            let last_observed = LastObserved {
+                block_height: checkpoint_height - blocks_rewound,
+                block_tx_idx: None,
+            };
+
+            // retain notes that correspond to transactions that are not "un-mined" after
+            // the rewind
+            let to_retain: BTreeSet<_> = self
+                .wallet_received_notes
+                .iter()
+                .filter_map(|(txid, n)| {
+                    if n.tx_height.iter().all(|h| h <= &last_observed.block_height) {
+                        Some(*txid)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+
+            self.spent_notes.retain(|_, v| to_retain.contains(&v.txid));
+            self.nullifiers.retain(|_, v| to_retain.contains(&v.txid));
+            self.wallet_received_notes
```

Comment:
In `zcashd` these transactions aren't evicted from the mempool (necessarily) so why are we always removing them here, when they can also represent notes that have only entered the mempool?